### PR TITLE
Add Shields filter list & "forget first party" metrics

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves https://github.com/brave/brave-browser/issues/37124

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

For `Brave.Shields.FilterLists`:

0. Start with fresh profile, ensure metric has value 0.
1. Access the Shields content filtering settings, enable one or two additional filter lists. Ensure the metric value is 1.
2. Add a custom filter list (i.e. https://ads-for-open-source.readthedocs.io/en/latest/_static/lists/opensource-ads.txt). Ensure the metric value is 3.
3. Disable the additional filter lists. Ensure the value is 2.
4. Remove the custom filter lists, ensure the value is 0.

For `Brave.Shields.ForgetFirstParty`:

0. Start with fresh profile, ensure metric has value 0.
1. Enable the "Forget me when I close this site" global Shields setting. Ensure metric has value 1.
2. Access any site, disable the "forget me" setting for the particular site. Ensure the metric has value 2.
3. Disable the global setting. Ensure the metric has value 0.
4. Access another site, enable the "forget me" setting for a particular site. Ensure the metric has value 3.

### CODE REVIEW

**Review by mkarolin — APPROVED** [review]
`chromium_src`++

**Review by iefremov — APPROVED** [review]

File: components/brave_shields/content/browser/ad_block_subscription_service_manager.h [hunk 1]
```diff
@@ -147,6 +150,8 @@ class AdBlockSubscriptionServiceManager {
   std::unique_ptr<component_updater::TimerUpdateScheduler>
       subscription_update_timer_ GUARDED_BY_CONTEXT(sequence_checker_);
 
+  raw_ptr<AdBlockListP3A> list_p3a_;
```
Comment (bridiver) [inline components/brave_shields/content/browser/ad_block_subscription_service_manager.h:153]: this should have GUARDED_BY_CONTENT to ensure it is only used on one thread, but I'm really not a fan of injecting P3A code directly into classes like this. I think we should have observer methods that P3A hooks into.

Comment (bridiver) [reply components/brave_shields/content/browser/ad_block_subscription_service_manager.h:153]: This class should also have an internal sequence checker to ensure that it is only ever called on the thread it is created on (it uses PrefService so can't be called on another thread. Alternatively you could get the current task runner when it is created and then always post back to the original thread so it is thread safe

File: components/brave_shields/core/browser/ad_block_component_service_manager.cc [hunk 2]
```diff
@@ -159,12 +162,15 @@ void AdBlockComponentServiceManager::UpdateFilterListPrefs(
...
-  update->Set(uuid, std::move(regional_filter_dict));
+  {
+    ScopedDictPrefUpdate update(local_state_, prefs::kAdBlockRegionalFilters);
+    base::Value::Dict regional_filter_dict;
+    regional_filter_dict.Set("enabled", enabled);
+    update->Set(uuid, std::move(regional_filter_dict));
+  }
 
   RecordP3ACookieListEnabled();
+  list_p3a_->ReportFilterListUsage();
```
Comment (bridiver) [inline components/brave_shields/core/browser/ad_block_component_service_manager.cc:173]: why do we have two separate p3a calls here done in two completely different ways? Can't we consolidate these into  an observer method and not include the p3a code in here at all? See my other comment about injecting P3A into everything which doesn't seem great to me

Comment (bridiver) [reply components/brave_shields/core/browser/ad_block_component_service_manager.cc:173]: the `UMA` stuff is inside the code everywhere, but at least it's always called in a consistent way

File: components/brave_shields/core/browser/ad_block_list_p3a.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+class PrefService;
+
+namespace brave_shields {
+
+class FilterListCatalogEntry;
+
+inline constexpr char kFilterListUsageHistogramName[] =
+    "Brave.Shields.FilterLists";
+
+class AdBlockListP3A {
```
Comment (bridiver) [inline components/brave_shields/core/browser/ad_block_list_p3a.h:23]: as mentioned above, this class needs a sequence checker

**Pull request merged**